lib/HLSL/HLMatrixSubscriptUseReplacer.h

```diff
@@ -0,0 +1,67 @@
+///////////////////////////////////////////////////////////////////////////////
+//                                                                           //
+// HLMatrixSubscriptUseReplacer.h                                            //
+// Copyright (C) Microsoft Corporation. All rights reserved.                 //
+// This file is distributed under the University of Illinois Open Source     //
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/IR/IRBuilder.h"
+#include <vector>
+
+namespace llvm {
+class Value;
+class AllocaInst;
+class CallInst;
+class Instruction;
+class Function;
+} // namespace llvm
+
+namespace hlsl {
+// Implements recursive replacement of a matrix subscript's uses,
+// from a pointer to a matrix value to a pointer to its lowered vector version,
+// whether directly or through GEPs in the case of two-level indexing like mat[i][j].
+// This has to handle one or two levels of indices, each of which either
+// constant or dynamic: mat[0], mat[i], mat[0][0], mat[i][0], mat[0][j], mat[i][j],
+// plus the equivalent element accesses: mat._11, mat._11_12, mat._11_12[0], mat._11_12[i]
+class HLMatrixSubscriptUseReplacer {
```

Comment:
BTW I'm thinking this could become a base class with virtual `cacheLoweredMatrix`/`loadElem`/`storeElem` functions, that could have an implementation for `HLMatrixLower` as is currently the case, and a second one for `HLOperationLower`, that would handle lowering subscripts to matrices living in constant/structured buffers. That would eliminate a lot of code duplication. Not in this PR though.